File: leetcode/121. Best Time to Buy and Sell Stock.cpp
// Problem link: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
// Time Complexity: O(n)
// Space Complexity: O(1)

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        int l = 0, r = 1, mp = 0;
        while (r < n) {
            int cp = prices[r] - prices[l];
            if (prices[l] < prices[r]) mp = max(mp, cp);
            else l = r;
            r++;
        }
        return mp;
    }
};

File: neetcode-150/sliding-window/121. Best Time to Buy and Sell Stock.cs
public class Solution {
    public int MaxProfit(int[] prices) {
        int n = prices.Length;
        int l = 0;
        int r = l + 1;
        int mp = 0;
        while (r < n)
        {
            int cp = prices[r] - prices[l];
            if (prices[r] > prices[l])
            {
                mp = Math.Max(mp, cp);
            }
            else l = r;
            r++;
        }
        return mp;
    }
}